File: 21_openspec_practise/examples/ecommerce-mini/src/services/order.js
export class OrderService {
  constructor(cartRepo, orderRepo, productRepo) {
    this.cartRepo = cartRepo
    this.orderRepo = orderRepo
    this.productRepo = productRepo
  }

  createOrder(userId) {
    const cart = this.cartRepo.findByUserId(userId)
    if (!cart || cart.items.length === 0) {
      throw new Error('CART_EMPTY')
    }

    // 1. Validate stock and calculate total
    let totalCents = 0
    const orderItems = []

    for (const item of cart.items) {
      const product = this.productRepo.findById(item.productId)
      if (!product) throw new Error(`Product ${item.productId} not found`)
      if (product.stock < item.quantity) throw new Error('OUT_OF_STOCK')
      
      totalCents += product.priceCents * item.quantity
      orderItems.push({
        productId: item.productId,
        priceCents: product.priceCents,
        quantity: item.quantity
      })
    }

    // 2. Deduct stock (Simulated Transaction)
    for (const item of cart.items) {
      const product = this.productRepo.findById(item.productId)
      product.stock -= item.quantity
      this.productRepo.save(product)
    }

    // 3. Create Order
    const order = {
      id: `order_${Math.random().toString(36).substr(2, 9)}`,
      userId,
      status: 'PENDING_PAYMENT',
      totalCents,
      items: orderItems
    }
    this.orderRepo.save(order)

    // 4. Clear Cart
    this.cartRepo.save({ userId, items: [] })

    return order
  }

  listOrders(userId) {
    if (!userId) throw new Error('MISSING_USER_ID')
    return this.orderRepo.findAll().filter(o => o.userId === userId)
  }
}

File: 21_openspec_practise/examples/ecommerce-mini-python/src/services/order.py
import uuid
from typing import List
from ..domain.models import Order, OrderItem
from ..repo.memory import MemoryRepo
from .cart import CartService
from .catalog import CatalogService

class OrderService:
    def __init__(self, repo: MemoryRepo[Order], cart_svc: CartService, catalog_svc: CatalogService):
        self.repo = repo
        self.cart_svc = cart_svc
        self.catalog = catalog_svc

    def create_order(self, user_id: str) -> Order:
        # 1. Get Cart
        cart = self.cart_svc.get_cart(user_id)
        if not cart.items:
            raise ValueError("CART_EMPTY")

        order_items = []
        total_cents = 0

        # 2. Validate and Calculate
        # Note: Ideally should lock resources. This is a simplified sequential check.
        for item in cart.items:
            product = self.catalog.get_product(item.product_id)
            if not product:
                raise ValueError(f"Product {item.product_id} not found")
            if product.stock < item.quantity:
                raise ValueError("OUT_OF_STOCK")
            
            total_cents += product.price_cents * item.quantity
            order_items.append(OrderItem(
                productId=item.product_id,
                priceCents=product.price_cents,
                quantity=item.quantity
            ))

        # 3. Deduct Stock
        for item in cart.items:
            product = self.catalog.get_product(item.product_id)
            product.stock -= item.quantity
            # Since objects are mutable references in memory repo, save is implicit but good practice
            self.catalog.repo.save(product.id, product)

        # 4. Create Order
        order = Order(
            id=f"order_{uuid.uuid4().hex[:8]}",
            userId=user_id,
            status="PENDING_PAYMENT",
            totalCents=total_cents,
            items=order_items
        )
        self.repo.save(order.id, order)

        # 5. Clear Cart
        self.cart_svc.clear_cart(user_id)

        return order

    def list_orders(self, user_id: str) -> List[Order]:
        if not user_id:
            raise ValueError("MISSING_USER_ID")
        return [o for o in self.repo.find_all() if o.user_id == user_id]
